internal/tofu/evaluate.go

```diff
@@ -352,9 +357,33 @@ func (d *evaluationStateData) GetLocalValue(addr addrs.LocalValue, rng tfdiags.S
 	return val, diags
 }
 
+func isCallFromRemote(parentCfg *configs.Config, moduleCall addrs.ModuleCall) bool {
+	if _, ok := parentCfg.SourceAddr.(addrs.ModuleSourceRemote); ok {
+		return true
+	}
+	calledModuleName := moduleCall.Name
+	parent := parentCfg.Parent
+	for parent != nil {
+		refCallCfg, ok := parent.Module.ModuleCalls[calledModuleName]
+		if !ok {
+			log.Printf("[ERROR] no module call found in %q for %q. This should not happen. Please report this to OpenTofu", parent.Path, calledModuleName)
```

Comment:
We don't typically expect end-users to notice messages in the logs, so if we want folks to report this as a bug then it should probably be either a normal error diagnostic or a panic.

However, we do also ask folks to send us the trace logs when they report a strange behavior, so it would be a reasonable compromise to just remove the "This should not happen..." part of this message and rework this to be something we can notice in the trace logs that folks send us:

```
[ERROR] tofu.isCallFromRemote: no module call found in foo for bar
```